### PR TITLE
Ability to set gunicorn options for API services; exponential backoff for clients

### LINKED ISSUE #111: Make `ComputeService`s tolerant of down `ComputeAPIService`
Currently, a `ComputeService` that encounters a down `ComputeAPIService` (due to timeouts on requests) will raise an exception and exit. Instead, we would like the `ComputeService` to retry with an exponential backoff, eventually giving up after some configurable number of attempts.

We can implement this by:
- adding a decorator that does retries with exponential backoff, eventual give-up configurable as parameter
- using decorator to wrap methods that call out via client

### PR DESCRIPTION
**Pull request by dotsdl**
Fixes #111.

### CODE REVIEW

**Comment by dotsdl** [comment]
After much gnashing of teeth, this is now ready for review! Both the user-facing and compute clients are now very tolerant of unreachable API services/timeouts, and we are now able to set gunicorn options server-side, such as longer timeouts.

I've performed burn tests on our development instance both user-facing and compute-side, and the API services appear to work as expected with distributing work among workers. 

We do observe a bottleneck compute-side with task claiming when there are a small number of `TaskHub`s/`AlchemicalNetwork`s with available `Task`s, since the claiming code must currently lock the `TaskHub` while performing claiming. This linearizes requests for new `Task`s from compute services, causing a bit of starving compute.

This can be mitigated by setting the number of claimed tasks on compute services higher. The problem will also be diminished when many networks have waiting tasks, as well as when tasks take much longer than the ~1min tasks I've been testing with.

@hmacdope and/or @mikemhenry, please give this a look when you get the chance! We should finally be clear to cut 0.1.0 and get our first users up and running!

**Comment by hmacdope** [comment]
Reviewing now! Excited for 0.1.0!

> We do observe a bottleneck compute-side with task claiming when there are a small number of `TaskHub`s/`AlchemicalNetwork`s with available `Task`s, since the claiming code must currently lock the `TaskHub` while performing claiming. This linearizes requests for new `Task`s from compute services, causing a bit of starving compute.

Should we raise an issue as a springboard for some improvement to this down the line?

**Comment by dotsdl** [comment]
@hmacdope yes! I'll make that issue with notes from my observations. I'm hopeful that under typical usage it isn't a huge issue, but under the conditions I noted it can be suboptimal compute-wise.

**Review by hmacdope — CHANGES_REQUESTED** [review]
Some minor comments. :) Other than that looks good!

File: .github/workflows/docker.yaml
Comment (dotsdl) [reply .github/workflows/docker.yaml:14]: We can remove this at any time after merge; it was only added to make image creation immediate on pushes to this branch. Same with `deployment-envs` before it.

File: alchemiscale/compute/api.py
Comment (dotsdl) [reply alchemiscale/compute/api.py:70]: To my understanding, we don't actually want to use `async` for API endpoints that don't feature `await`s on other `async` functions. Many of our API endpoints are not IO bound, but compute bound, since there is a lot of (de)serialization happening.

If we decide later to make some of these calls take advantage of `async/await`, then we would use the `async` keyword upfront on those specifically.

Comment (dotsdl) [reply alchemiscale/compute/api.py:70]: It wasn't clear to me if the perf bottleneck I was seeing on `gunicorn+uvicorn` was due to our use of `async` everywhere despite having very compute-bound API endpoints, so I went ahead and removed them.

File: alchemiscale/compute/service.py
Comment (dotsdl) [reply alchemiscale/compute/service.py]: Both! I've clarified this better now.

File: alchemiscale/base/client.py
Comment (dotsdl) [reply alchemiscale/base/client.py:114]: Ah, I actually don't know if there's a difference. I have actually never used `raise from`.

Comment (hmacdope) [reply alchemiscale/base/client.py:114]: If `raise` is working as you expect lets just go with that.

**Comment by dotsdl** [comment]
Think I've addressed all your suggestions @hmacdope! Anything else you think this needs?

**Review by hmacdope — APPROVED** [review]
Looks great @dotsdl. Thanks for fixing everything up (and so quickly!)

**Comment by hmacdope** [comment]
Feel free to merge on green CI.

**Comment by dotsdl** [comment]
Thanks @hmacdope! I may add some more logging statements to the `SynchronousComputeService`, but that's it! I'll merge after checks and I'm satisfied there.

**Pull request merged**
